src/libasr/asr_utils.cpp

```diff
@@ -713,6 +714,68 @@ ASR::asr_t* symbol_resolve_external_generic_procedure_without_eval(
     }
 }
 
+ASR::asr_t* make_ImplicitCast_t_value(Allocator &al, const Location &a_loc, ASR::expr_t* a_arg, ASR::cast_kindType a_kind, ASR::ttype_t* a_type, ASR::expr_t* value) {
```

Comment:
Wouldn't a better API be to remove the `value` argument, and the idea being that you call this function if you want to compute the `value` automatically? But if you know `value` ahead of time, then you simply call the `make_ImplicitCast_t` directly.

Reply:
The reason I included the `value` argument is because now we can call only the `make_ImplicitCast_t_value()` and if there is a value, it can be included, and if the value is not provided, it will be `nullptr` by default, and the function will calculate the value of `value`.

So now we can stop using `make_ImplicitCast_t()` completely and just use `make_ImplicitCast_t_value()` for all purposes, since it will again call `make_ImplicitCast_t()` inside it.

If this is bad I will change it like you said.

Reply:
@certik I made a commit to the second suggestion you made. I will wait until you confirm this one again to make the other necessary changes.

Reply:
I see. It seems that you would almost always call this with "value=nullptr" as the argument, wouldn't you? In that case, I would just remove the argument altogether.

Reply:
So, there are two cases,

1. `value` arg is not `nullptr` - In this case you are just creating the `ImplicitCast` node and returning it.
2. `value` arg is `nullptr` - In this case you are doing the computation.

The second case makes sense. Is there any example when `value` is not `nullptr`. Because if its not `nullptr` then its already computed for `ImplicitCast` node and if that's the case then why should we even call this function?

So, the bottom line is if you know that `value` is not `nullptr` then directly call, `make_ImplicitCast_t` and at places where you are passing `nullptr` as `value` argument of your function just don't do that and instead update the API here to not accept any `value` parameter. That should do the job.

Reply:
I see the design now and have corrected it in the new commit. Thank you @certik and @czgdp1807 for explaining.